### PR TITLE
287580 - Datatable sort

### PR DESCRIPTION
**Pull request by MBausson**

### CODE REVIEW

**Comment by B-Rass** [comment]
@MBausson can you look after the conflicts? Thx!

File: app/controllers/modeles_controller.rb [hunk 1]
```diff
@@ -7,7 +7,7 @@ class ModelesController < ApplicationController
 
   def index
     @modeles = Modele.includes(:category, :enclosures).order(:name)
-    @types = @modeles.group_by { |m| m.category.name }.sort_by { |categorie, modeles| categorie.to_s }
+    @types = sorted @modeles.group_by { |m| m.category.name }.sort_by { |categorie, modeles| categorie.to_s }
```
Comment (nicolas-brousse) [inline app/controllers/modeles_controller.rb:10]: Doesn't work on this index. `sorted` is not on the right place.

File: app/views/architectures/index.html.erb [hunk 2]
```diff
@@ -23,40 +23,27 @@
...
-          </thead>
+      <%= render List::DataTableComponent.new(@architectures) do |table| %>
+        <% table.with_column(Architecture.human_attribute_name(:name), sort_by: :name) do |architecture| %>
+          <b>
+            <%= link_to architecture, architecture_path(architecture) %>
+          </b>
+        <% end %>
+
+        <% table.with_column(Architecture.human_attribute_name(:usage)) do |architecture| %>
+          <%= link_to pluralize(architecture.modeles.count, 'modèle', 'modèles'), modeles_path %>
```
Comment (nicolas-brousse) [inline app/views/architectures/index.html.erb:34]: Use i18n here. `Modele.model_name.human.pluralize(count: architecture.modeles.count)`.

File: app/views/sites/index.html.erb [hunk 3]
```diff
@@ -24,42 +24,30 @@
...
-                          site,
-                          method: :delete,
-                          data: { confirm: t(".delete_confirmation") },
-                          class: 'btn btn-danger' %>
-            </td>
-          <% end %>
-          </tbody>
-        </table>
-      </div>
+        <% table.with_column(Site.human_attribute_name(:position)) do |site| %>
```
Comment (nicolas-brousse) [inline app/views/sites/index.html.erb:34]:
```suggestion
        <% table.with_column(Site.human_attribute_name(:position), sort_by: :position) do |site| %>
```

File: app/views/servers/index.html.erb [hunk 4]
```diff
@@ -21,60 +21,43 @@
...
+        <% table.with_column(Server.human_attribute_name(:islet)) do |server| %>
+          <%= link_to server.islet, islet_path(server.islet) if server.islet %>
+        <% end %>
+
+        <% table.with_column(Server.human_attribute_name(:bay)) do |server| %>
+          <%= link_to server.bay, bay_path(server.bay) if server.bay %>
+        <% end %>
+
+        <% table.with_column(class: "text-right") do |server| %>
+          <div class="btn-group btn-group-sm btn-group-vertical" role="group" aria-label="...">
```
Comment (nicolas-brousse) [inline app/views/servers/index.html.erb:50]: Div to be removed. Let do the same as other tables. We will do some more changes later if necessary.

File: app/views/rooms/index.html.erb [hunk 5]
```diff
@@ -18,47 +18,41 @@
...
+          <input type="checkbox" <%= room.display_on_home_page ? "checked" : "" %> disabled>
+        <% end %>
+
+        <% table.with_column(class: "text-right") do |room| %>
+          <%= render partial: "rooms/export_button", locals: { room: room } %>
+          <%= link_to 'Modifier', edit_room_path(room), class: 'btn btn-primary' %>
+          <%= link_to 'Supprimer',
+                      room,
+                      method: :delete,
+                      data: { confirm: 'Cette salle ne pourra pas être supprimée tant qu\'elle sera associée à au moins un ilôt. Voulez-vous continuer ?' },
```
Comment (nicolas-brousse) [inline app/views/rooms/index.html.erb:52]: Move to i18n

File: app/views/rooms/index.html.erb [hunk 6]
```diff
@@ -18,47 +18,41 @@
...
+          <%= room.position %>
+        <% end %>
+
+        <% table.with_column("Contenu") do |room| %>
+          <%= pluralize room.islets.count, 'îlot', 'îlots' %>
+          /
+          <%= pluralize room.frames.count, 'chassis' %>
+        <% end %>
+
+        <% table.with_column("Afficher sur la page d'accueil", sort_by: :display_on_home_page) do |room| %>
```
Comment (nicolas-brousse) [inline app/views/rooms/index.html.erb:42]: Move to i18n

File: app/views/rooms/index.html.erb [hunk 7]
```diff
@@ -18,47 +18,41 @@
...
+        <% end %>
+
+        <% table.with_column("Position", sort_by: :position) do |room| %>
+          <%= room.position %>
+        <% end %>
+
+        <% table.with_column("Contenu") do |room| %>
+          <%= pluralize room.islets.count, 'îlot', 'îlots' %>
+          /
+          <%= pluralize room.frames.count, 'chassis' %>
```
Comment (nicolas-brousse) [inline app/views/rooms/index.html.erb:39]: Move to i18n

File: app/views/rooms/index.html.erb [hunk 8]
```diff
@@ -18,47 +18,41 @@
...
+            <%= link_to room, room_path(room) %>
+          </b>
+        <% end %>
+
+        <% table.with_column("Position", sort_by: :position) do |room| %>
+          <%= room.position %>
+        <% end %>
+
+        <% table.with_column("Contenu") do |room| %>
+          <%= pluralize room.islets.count, 'îlot', 'îlots' %>
```
Comment (nicolas-brousse) [inline app/views/rooms/index.html.erb:37]: Move to i18n

File: app/views/rooms/index.html.erb [hunk 9]
```diff
@@ -18,47 +18,41 @@
...
-                          room,
-                          method: :delete,
-                          data: { confirm: 'Cette salle ne pourra pas être supprimée tant qu\'elle sera associée à au moins un ilôt. Voulez-vous continuer ?' },
-                          class: 'btn btn-danger' %>
-            </td>
-          <% end %>
-          </tbody>
-        </table>
-      </div>
+        <% table.with_column("Nom", sort_by: :name) do |room| %>
```
Comment (nicolas-brousse) [inline app/views/rooms/index.html.erb:26]: Move to i18n

File: app/views/rooms/index.html.erb [hunk 10]
```diff
@@ -18,47 +18,41 @@
...
+          <b>
+            <%= link_to room, room_path(room) %>
+          </b>
+        <% end %>
+
+        <% table.with_column("Position", sort_by: :position) do |room| %>
```
Comment (nicolas-brousse) [inline app/views/rooms/index.html.erb:32]: Move to i18n

File: app/views/rooms/index.html.erb [hunk 11]
```diff
@@ -18,47 +18,41 @@
...
+          <%= room.position %>
+        <% end %>
+
+        <% table.with_column("Contenu") do |room| %>
```
Comment (nicolas-brousse) [inline app/views/rooms/index.html.erb:36]: Move to i18n

File: app/views/rooms/index.html.erb [hunk 12]
```diff
@@ -18,47 +18,41 @@
...
-            <th>Site</th>
-            <th>Nom</th>
-            <th>Position</th>
-            <th>Contenu</th>
-            <th>Afficher sur la page d'accueil</th>
-            <th></th>
-          </tr>
-          </thead>
+      <%= render List::DataTableComponent.new(@rooms) do |table| %>
+        <% table.with_column("Site") do |room| %>
```
Comment (nicolas-brousse) [inline app/views/rooms/index.html.erb:22]: Move to i18n

**Review by nicolas-brousse — CHANGES_REQUESTED** [review]

File: app/views/categories/index.html.erb [hunk 13]
```diff
@@ -23,38 +23,26 @@
...
-            </td>
-          <% end %>
-          </tbody>
-        </table>
-      </div>
+      <%= render List::DataTableComponent.new(@categories) do |table| %>
+        <% table.with_column(Category.human_attribute_name(:name), sort_by: :name) do |category| %>
+          <b>
+            <%= link_to category, category_path(category) %>
+          </b>
```
Comment (nicolas-brousse) [inline app/views/categories/index.html.erb:30]: Is it for a particular reason that you put a `<b>` tag around?

File: app/views/categories/index.html.erb [hunk 14]
```diff
@@ -23,38 +23,26 @@
...
+        <% end %>
+
+        <% table.with_column(Category.human_attribute_name(:modeles_count)) do |category| %>
+          <%= link_to "#{pluralize(category.modeles.count, 'modèle', 'modèles')}", modeles_path %>
```
Comment (nicolas-brousse) [inline app/views/categories/index.html.erb:34]: Use i18n.

File: app/views/contract_types/index.html.erb [hunk 15]
```diff
@@ -23,34 +23,21 @@
...
+        <% table.with_column(ContractType.human_attribute_name(:name), sort_by: :name) do |contract_type| %>
+          <%= link_to contract_type, contract_type_path(contract_type) %>
+        <% end %>
+
+        <% table.with_column(class: "text-right") do |contract_type| %>
+          <%= link_to t("action.edit"), edit_contract_type_path(contract_type), class: 'btn btn-primary' %>
+          <%= link_to t("action.delete"),
+                      contract_type,
+                      method: :delete,
+                      data: { confirm: 'Toutes les données concernant ce type de contrat seront supprimées, ainsi que les contrats de maintenance associés. Voulez-vous continuer ?' },
```
Comment (nicolas-brousse) [inline app/views/contract_types/index.html.erb:36]: Moe in i18n.

File: app/views/disk_types/index.html.erb [hunk 16]
```diff
@@ -23,30 +23,19 @@
...
-              </td>
-            <% end %>
-          </tbody>
-        </table>
-      </div>
+      <%= render List::DataTableComponent.new(@disk_types) do |table| %>
+        <% table.with_column(DiskType.human_attribute_name(:name), sort_by: :name) do |disk_type| %>
+          <b>
+            <%= link_to disk_type, disk_type_path(disk_type) %>
+          </b>
```
Comment (nicolas-brousse) [inline app/views/disk_types/index.html.erb:30]: Is it for a particular reason that you put a `<b>` tag around?

File: app/views/domaines/index.html.erb [hunk 17]
```diff
@@ -23,40 +23,26 @@
...
-              <%= Domaine.human_attribute_name(:usage) %>
-            </th>
-            <th></th>
-          </tr>
-          </thead>
+      <%= render List::DataTableComponent.new(@domaines) do |table| %>
+        <% table.with_column(Domaine.human_attribute_name(:name), sort_by: :name) do |domaine| %>
+          <b>
+            <%= link_to domaine, domaine_path(domaine) %>
+          </b>
```
Comment (nicolas-brousse) [inline app/views/domaines/index.html.erb:30]: Is it for a particular reason that you put a <b> tag around?

File: app/views/islets/index.html.erb [hunk 18]
```diff
@@ -23,55 +23,39 @@
...
-        </table>
-      </div>
+        <% table.with_column(Room.model_name.human) do |islet| %>
+          <%= link_to islet.room, room_path(islet.room) %>
+        <% end %>
+
+        <% table.with_column(Islet.model_name.human, sort_by: :name) do |islet| %>
+          <b>
+            <%= link_to islet, islet_path(islet) %>
+          </b>
```
Comment (nicolas-brousse) [inline app/views/islets/index.html.erb:38]: Is it for a particular reason that you put a <b> tag around?

File: app/views/maintainers/index.html.erb [hunk 19]
```diff
@@ -22,34 +22,21 @@
...
+        <% table.with_column(class: "text-right") do |maintainer| %>
+          <%= link_to t("action.edit"), edit_maintainer_path(maintainer), class: 'btn btn-primary' %>
+          <%= link_to t("action.delete"),
+                      maintainer,
+                      method: :delete,
+                      data: { confirm: t(".delete_confirmation") },
+                      class: 'btn btn-danger' %>
+        <% end %>
+      <% end %>
+
```
Comment (nicolas-brousse) [inline app/views/maintainers/index.html.erb:39]:
```suggestion
```

Remove empty line

**Review by nicolas-brousse — CHANGES_REQUESTED** [review]

File: app/views/categories/index.html.erb (hunk 13 above)
Comment (MBausson) [reply app/views/categories/index.html.erb:30]: I think I've seen it somewhere else before, I added it here so it is consistent

File: app/controllers/modeles_controller.rb [hunk 20]
```diff
@@ -7,7 +7,7 @@ class ModelesController < ApplicationController
 
   def index
     @modeles = Modele.includes(:category, :enclosures).order(:name)
-    @types = @modeles.group_by { |m| m.category.name }.sort_by { |categorie, modeles| categorie.to_s }
+    @types = sorted(@modeles).group_by { |m| m.category.name }.sort_by { |categorie, modeles| categorie.to_s }
```
Comment (nicolas-brousse) [inline app/controllers/modeles_controller.rb:10]: Prefer to call sorted on line 9 than here

**Review by nicolas-brousse — CHANGES_REQUESTED** [review]

File: config/locales/activerecord.fr.yml [hunk 21]
```diff
@@ -41,6 +41,17 @@ fr:
       changelog_entry: Changelog
 
     attributes:
+      room:
+        name: Nom
+        position: Position
+        content: Contenu
+        display_on_home_page: Afficher sur la page d'accueil
+        islets_count:
+          one: 1 îlot
```
Comment (B-Rass) [inline config/locales/activerecord.fr.yml:50]:
```suggestion
         zero: 0 îlot
          one: 1 îlot
```

File: config/locales/activerecord.fr.yml [hunk 22]
```diff
@@ -41,6 +41,17 @@ fr:
...
+          other: '%{count} îlots'
+        frames_count:
+          one: 1 châssis
```
Comment (B-Rass) [inline config/locales/activerecord.fr.yml:53]:
```suggestion
          zero: 0 châssis
          one: 1 châssis
```

File: config/locales/activerecord.fr.yml [hunk 23]
```diff
@@ -180,7 +191,9 @@ fr:
       category:
         name: Nom
         description: Description
-        modeles_count: Usage
+        modeles_count:
+          one: 1 modèle
```
Comment (B-Rass) [inline config/locales/activerecord.fr.yml:195]:
```suggestion
          zero: 0 modèle
          one: 1 modèle
```

**Review by B-Rass — CHANGES_REQUESTED** [review]
3 cleans sur les clés de trad'

si on mets pas la clé `zero` alors ça va afficher `1 xxxx`, même si le `count` = 0

File: app/views/islets/index.html.erb (hunk 18 above)
Comment (nicolas-brousse) [reply app/views/islets/index.html.erb]: @MBausson ?

File: app/views/categories/index.html.erb (hunk 13 above)
Comment (nicolas-brousse) [reply app/views/categories/index.html.erb]: Ok for me. But in this case make sure to add it on every table. It seems it is not always present.

**Review by nicolas-brousse — APPROVED** [review]

File: app/views/architectures/index.html.erb [hunk 24]
```diff
@@ -23,42 +20,26 @@
...
-          </tbody>
-        </table>
-      </div>
-    </div>
+        <% table.with_column(Architecture.human_attribute_name(:usage)) do |architecture| %>
+          <%= link_to Architecture.human_attribute_name(:modeles_count, count: architecture.modeles.count), modeles_path %>
+        <% end %>
 
+        <% table.with_column(class: "text-right") do |architecture| %>
+          <%= link_to 'Modifier', edit_architecture_path(architecture), class: 'btn btn-primary' %>
```
Comment (B-Rass) [inline app/views/architectures/index.html.erb:35]: Use traduction key here

File: app/views/architectures/index.html.erb [hunk 25]
```diff
@@ -23,42 +20,26 @@
...
+          <%= link_to 'Supprimer',
```
Comment (B-Rass) [inline app/views/architectures/index.html.erb:36]: Use traduction key here

**Review by B-Rass — CHANGES_REQUESTED** [review]

File: app/controllers/disk_types_controller.rb [hunk 26]
```diff
@@ -6,7 +6,7 @@ class DiskTypesController < ApplicationController
   # GET /disk_types
   # GET /disk_types.json
   def index
-    @disk_types = DiskType.all
+    @disk_types = DiskType.order(unit: :desc, quantity: :desc)
```
Comment (B-Rass) [inline app/controllers/disk_types_controller.rb:9]:
```suggestion
    @disk_types = DiskType.order(unit: :asc, quantity: :asc)
```

**Review by B-Rass — CHANGES_REQUESTED** [review]

File: app/controllers/disk_types_controller.rb (hunk 26 above)
Comment (nicolas-brousse) [reply app/controllers/disk_types_controller.rb]: Done

File: app/views/architectures/index.html.erb (hunk 25 above)
Comment (nicolas-brousse) [reply app/views/architectures/index.html.erb]: Done

File: app/views/architectures/index.html.erb (hunk 24 above)
Comment (nicolas-brousse) [reply app/views/architectures/index.html.erb]: Done

**Review by B-Rass — APPROVED** [review]

**Pull request merged**